kafka-transport-binding.md

```diff
@@ -0,0 +1,271 @@
+of its choosing, but as the value is transposed into Kafka as defined in this
+specification, core Kafka provides data available as a sequence of bytes.
+
+For instance, if the declared `contentType` is
+`application/json;charset=utf-8`, the expectation is that the `data` attribute
+value is made available as [UTF-8][RFC3629] encoded JSON text.
+
+## 3. Kafka Message Mapping
+
+With Kafka 0.11.0.0 and above, the content mode is chosen by the sender of the 
+event. Protocol usage patterns that might allow solicitation of events using a
+particular content mode might be defined by an application, but are not defined
+here.
+
+The receiver of the event can distinguish between the two content modes by
+inspecting the `cloudEvents_contentType` property of the Kafka message. If the
+value is prefixed with the CloudEvents media type `application/cloudevents`, 
+indicating the use of a known [event format](#14-event-formats), the receiver
+uses *structured* mode, otherwise it defaults to *binary* mode.
+
+If a receiver finds a CloudEvents media type as per the above rule, but with an
+event format that it cannot handle, for instance
+`application/cloudevents+avro`, it MAY still treat the event as binary and
+forward it to another party as-is.
+
+### 3.1. Binary Content Mode
+
+The *binary* content mode accommodates any shape of event data, and allows for
+efficient transfer and without transcoding effort.
+
+#### 3.1.1. Content Type
+
+For the *binary* mode, the header `cloudEvents_contentType` property MUST be
+mapped directly to the CloudEvents `contentType` attribute.
+
+#### 3.1.2. Event Data Encoding
+
+The [`data` attribute](#22-data-attribute) byte-sequence MUST be used as the
+value of the Kafka message.
+
+#### 3.1.3. Metadata Headers
+
+All [CloudEvents][CE] attributes and
+[CloudEvent Attributes Extensions](primer.md#cloudevent-attribute-extensions) 
+with exception of `data` MUST be individually mapped to and from the Header 
+fields in the Kafka message.
+
+##### 3.1.3.1 Property Names
+
+Cloud Event attributes are prefixed with "cloudEvents_" for use in the
+[message-headers][Kafka-Message-Header] section.
+
+Examples:
+
+    * `eventTime` maps to `cloudEvents_eventTime`
+    * `eventID` maps to `cloudEvents_eventID`
+    * `cloudEventsVersion` maps to `cloudEvents_cloudEventsVersion`
+
+##### 3.1.3.2 Property Values
+
+The value for each Kafka header is constructed from the respective
+header's Kafka representation, compliant with the [Kafka message
+format][Kafka-Message-Format] specification.
+
+
+#### 3.1.4 Example
+
+This example shows the *binary* mode mapping of an event into the
+Kafka message. All other CloudEvents attributes
+are mapped to Kafka Header fields with prefix `cloudEvents_`.
+
+Mind that `cloudEvents_` here does refer to the event `data`
+content carried in the payload.
+
+``` text
+------------------ Message -------------------
+
+Topic Name: mytopic
+
+------------------- key ----------------------
+
+Key: mykey
```

Comment:
The Kafka message key is where things are still a bit blurry to me. How will its value be obtained from the CloudEvents event? Have you foreseen some extractor function or similar?

Reply:
Thinking more about that one, perhaps a header key should be defined whose value, if present, will be propagated as key of the Kafka message, e.g. `cloudEvents_messageKey`. Alternatively, the binding, wherever it's running, could be configured with some kind of path expression which gets applied to the `data` value to extract the value, e.g. `/customer/id`.

Reply:
Ah yes, thanks. I've commented over there, too. Seems that'd need resolution first.

Reply:
The conflict we have here is that the partition key is a transport specific concern (and the requirements may vary across different transport options) and that the publisher ought not to care about what transport the event route gets bound to. We may also have the situation that an event gets first published via MQTT from a little device and then gets put on Kafka by a device gateway. 

I think this binding needs to define a rule by which a key is constructed from the event rather than expecting that the event brings it along.

Reply:
I'm not sure if `source` is the right place - I'd expect all events produced by same application to have identical source, why partitioning typically applies to events produced by same source. 

I like the proposal from @clemensv in #218 : 
" instead of putting the burden of producing an event key on the client and then only having one, any transport that requires particular constructs such as keys such define a mechanism by which you can harvest/synthesize a key from an incoming CloudEvent as some sort of transform. The spec doesn't need to prescribe how -- it just needs to say that that's how the key materializes. A transform that just cooks up a random key might also be valid if that's what you want."

KafkaConnect already has "key extraction" transformation, exactly because external records require mapping to Kafka keys, and the logic for doing so varies between use-cases.

Reply:
@gwenshap @clemensv that would mean that the kafka transport spec would leave the determination of the "key" as an exercise for the implementer (or admin configuring the transport) ?

Reply:
Just added a comment on this over at https://github.com/cloudevents/spec/pull/218. The idea being that instead of having fully pluggable logic for retrieving the key instead there could be a mechanism to just select the key from a given event field when sending a cloud event to a Kafka Topic.